docs/compilers/terrapin.md

```diff
@@ -0,0 +1,125 @@
+- Validation of DLLs on Windows Part 2: the target of this milestone is NuPkgs produced on Windows using the .NET Core compiler. This will require Terrapin to take the extra environment setup step of installing the correct .NET Core runtime and using that to drive the installation process.
+- Validate binaries produced on Linux / Mac: the target of this milestone is NuPkgs produced on Linux / Mac. This will require Terrapin to involve multiple machines in the validation process. It will also help us understand the true impact of OS sub-versions in our ability to do provenance validation.
+
+### FAQ
+
+### What about binaries produced before VS 16.7 was released?
+The provenance validation process depends on extra metadata included in the PDB that the compiler uses to generate the artifacts manifest file as well as recreate the compilation. That extra information was first added into PDBs in the Visual Studio 16.7 release. That means any binaries produced before then will fail provenance validation. 
+
+This is a known issue and the understanding is such older NuPkgs will not be supported for provenance validation. 
+
+### How do we handle source generators?
+The provenance validation will occur using the output of source generators rather than re-running source generators as part of the validation. This will simplify the process as well as remove the issue of determinism in source generators (customer defined tools that don't have our level of rigor) from the process.
+
+## Open Issues
+
+### Running the validator on .NET Framework
+The expectation is that for Windows we should be able to use .NET Core to run the validator for binaries that were produced using the .NET Framework compiler. The runtime differences we are worried about for Unix operating systems don't have the same problems on Windows. Hence any .NET Runtime which can execute the compiler should be able to do provenance validation. 
+
+It's possible we will find differences here though in the real world. In which case we will need to have a mode for `dotnet-roslyn-validator` where by it will shell out to a .NET Framework process to do the validation. 
+
+### Line ending issues
+In order for validation to succeed it is important for `dotnet-roslyn-validator` to receive source files that have the same line endings as the one's provided to the original compilation. These factor into items like the checksum of source files stored in PDBs, the content of string literals, etc ... 
+
+It's possible that the information provided by source link is not sufficient to properly recreate files with correct line endings. Consider that a `.gitattributes` file could change how line endings are used on different operating systems. 
+
+We need to see how much of a problem this is in practice before we decide how to approach it. One possible solution is as simple as giving customers guidance on how to handle line endings in a provenance friendly manner.
```

Comment:
The debugger (and Source Link validation tool) has exactly the same problem with source lines since when it retrieves a file from source control server from Source Link URL it checks whether the checksum stored in the PDB matches the hash of the content of the file. 

The solution is rather simple: first check the file as is. If it doesn't match detect the line endings. If the file has mixed line endings than fail the check. If the file has consistent LF endings replace them with CRLF endings and calculate the hash. If the hash doesn't match fail the check. Otherwise we now have the exact file content.

See https://github.com/dotnet/sourcelink/pull/678/files#diff-3339ce7022ba5de3029d3f15bdf8ca585905bad4f25077ca1b1616c31d841c62R258-R290

Reply:
That does seem like a good aproach we can take here.